python/cm/issue.py

```diff
@@ -10,18 +10,89 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from cm.logger import log   # pylint: disable=unused-import
+from collections import defaultdict
+from itertools import chain
+
 import cm.status_api
-from cm.errors import AdcmEx
-from cm.errors import raise_AdcmEx as err
 from cm.adcm_config import proto_ref, obj_ref, get_prototype_config
-from cm.models import ConfigLog, Host, ClusterObject, Prototype, HostComponent
-from cm.models import PrototypeImport, ClusterBind
+from cm.errors import AdcmEx, raise_AdcmEx as err
+from cm.logger import log
+from cm.models import (
+    ClusterBind,
+    ClusterObject,
+    ConfigLog,
+    Host,
+    HostComponent,
+    Prototype,
+    PrototypeImport,
+    ServiceComponent,
+)
+
+
+def update_hierarchy_issues(obj):
+    """
+    Cluster issue affects cluster's services and their components; also affects its hosts;
+    ClusterService issue affects its cluster and its components;
+    ServiceComponent issue affects its service and cluster of service;
+
+    HostProvider issue affects its hosts;
+    Host issue affects its cluster and its provider.
+
+    objects with unknown prototype are ignored
+    """
+    for item in chain(
+            get_affected_cluster_hierarchy(obj),
+            get_affected_host_hierarchy(obj)
+    ):
+        save_issue(item)
+
+
+def get_affected_cluster_hierarchy(obj) -> list:
+    result = []
+
+    if obj.prototype.type == 'cluster':
+        result.append(obj)
+        for co in ClusterObject.objects.filter(cluster=obj):
+            result.append(co)
+            result.extend(ServiceComponent.objects.filter(cluster=obj, service=co))
+
+    elif obj.prototype.type == 'service':
+        result.append(obj.cluster)
+        result.append(obj)
+        result.extend(ServiceComponent.objects.filter(service=obj))
+
+    elif obj.prototype.type == 'component':
+        result.append(obj.service.cluster)
+        result.append(obj.service)
+        result.append(obj)
+
+    elif obj.prototype.type == 'host':
+        result.append(obj.cluster)
```

Comment:
Don't we need to add
 
`result.append(obj)`

here?

Reply:
No,  it's in other function (get_affected_host_hierarchy), same for cluster there